src/Compiler/Microsoft.AspNetCore.Razor.Language/src/MetadataCollection.cs

```diff
@@ -231,47 +252,176 @@ private sealed class OneToThreeItems : MetadataCollection
+            if (key3 is null)
+            {
+                throw new ArgumentNullException(nameof(key3));
+            }
+
+            var compareKeys_1_2 = string.CompareOrdinal(key1, key2);
+
+            if (compareKeys_1_2 == 0)
             {
                 throw new ArgumentException(
                     Resources.FormatAn_item_with_the_same_key_has_already_been_added_Key_0(key2), nameof(key2));
             }
 
-            if (key1 == key3)
+            var compareKeys_1_3 = string.CompareOrdinal(key1, key3);
+
+            if (compareKeys_1_3 == 0)
             {
                 throw new ArgumentException(
                     Resources.FormatAn_item_with_the_same_key_has_already_been_added_Key_0(key3), nameof(key3));
             }
 
-            if (key2 == key3)
+            var compareKeys_2_3 = string.CompareOrdinal(key2, key3);
+
+            if (compareKeys_2_3 == 0)
             {
                 throw new ArgumentException(
                     Resources.FormatAn_item_with_the_same_key_has_already_been_added_Key_0(key3), nameof(key3));
             }
 
+            // We ensure that the key/value pairs are assigned to their fields in sorted order to ensure
+            // that MetadataCollections created with the same data in a different order are equal.
+
+            var pair1 = (key1, value1);
+            var pair2 = (key2, value2);
+            var pair3 = (key3, value3);
+
+            if (compareKeys_1_2 < 0)
+            {
+                if (compareKeys_1_3 < 0)
+                {
+                    // If key1 is less than both key2 and key3, it must go first.
+                    // - slot 1: pair 1
+                    (_key1, _value1) = pair1;
+
+                    // Now that key1 is handled, we can determine the other two slots by comparing key2 and key3.
+                    if (compareKeys_2_3 < 0)
+                    {
+                        // - slot 2: pair 2
+                        // - slot 3: pair 3
+                        (_key2, _value2) = pair2;
+                        (_key3, _value3) = pair3;
+                    }
+                    else
+                    {
+                        // - slot 2: pair 3
+                        // - slot 3: pair 2
+                        (_key2, _value2) = pair3;
+                        (_key3, _value3) = pair2;
+                    }
+                }
+                else
+                {
+                    // If key1 is less than key2 but not key3, it must go in the middle.
+                    // - slot 1: pair 3
+                    // - slot 2: pair 1
+                    // - slot 3: pair 2
+                    (_key1, _value1) = pair3;
+                    (_key2, _value2) = pair1;
+                    (_key3, _value3) = pair2;
+                }
+            }
+            else if (compareKeys_1_3 < 0)
+            {
+                // If key1 is less than key3 but not key2, it must go in the middle.
```

Comment:
Shouldn't this also imply that `key3` < `key2`?

Reply:
Yes, that's right. Tweaking...